yamcs-core/src/main/java/org/yamcs/tctm/ccsds/UdpTmFrameLink.java

```diff
@@ -63,24 +65,75 @@ public void doStop() {
     public void run() {
         while (isRunningAndEnabled()) {
             try {
-                tmSocket.receive(datagram);
-                if (log.isTraceEnabled()) {
-                    log.trace("Received datagram of length {}: {}", datagram.getLength(), StringConverter
-                            .arrayToHexString(datagram.getData(), datagram.getOffset(), datagram.getLength(), true));
+
+                // Detect if the Attached Synchro Marker (ASM) is present in the data link part of the yamcs.instance.yaml file
+                asmPresent = config.getBoolean("asmPresent", false); // By default ASM is setted as false   
+                
+                // Array to select the first four bytes 
+                byte[] firstsBytes = new byte[4];
+                
+                if (!asmPresent) {
+                    tmSocket.receive(datagram);
+
+                    // Select the first four bytes
+                    for(int i = 0; i < 4; i++){
+                        firstsBytes[i]=datagram.getData()[i];
+                    }
+                    String firstBytesStr = convertBytesToHexadecimal(firstsBytes);
+
+                    if (firstBytesStr.equals("1ACFFC1D"))
+                        throw new IllegalArgumentException("You specified your frame does not begin with the Attached Synchronization Marker but it seems it is...");
+
+                    else { // If !asmPresent and the data indeed does not start with the ASM 
+                        if (log.isTraceEnabled()) {
+                            log.trace("Received datagram of length {}: {}", datagram.getLength(), StringConverter
+                                    .arrayToHexString(datagram.getData(), datagram.getOffset(), datagram.getLength(), true));
+                        }
+                    }
+                }
+
+                else {
+                    int maxLength = frameHandler.getMaxFrameSize();
+                    DatagramPacket datagramWithAsm = new DatagramPacket(new byte[maxLength + 4 ], maxLength + 4);
+                    datagram = new DatagramPacket(new byte[maxLength], maxLength);
+                    tmSocket.receive(datagramWithAsm);
+                    
+                    // Select the first four bytes
+                    for(int i = 0; i < 4; i++){
+                        firstsBytes[i]=datagramWithAsm.getData()[i];
+                    }
+                    String firstBytesStr = convertBytesToHexadecimal(firstsBytes);
+                    System.out.println("ASM: "+firstBytesStr);
+
+                    if (!firstBytesStr.equals("1ACFFC1D")){
+                        throw new IllegalArgumentException("You specified your frame begins with the Attached Synchronization Marker word but it is not.");
+                    }
+
+                    else{ // If asmPresent and the data indeed start with the ASM       
```

Comment:
get rid of this else, just use asmLength in the code to send the frame

Reply:
I am not sure to understand this one. Would you like to make only one case, with a parameter asmLength which value would be either 0 or 4?

Reply:
I mean just have a asmLength = 0 when there is no ASM and have only one code path that deals with both ASM and non/ASM case.

And you can call this:
 handleFrame(timeService.getHresMissionTime(), datagram.getData(), datagram.getOffset() + asmLength,
                        datagram.getLength() - asmLength);

Reply:
Yes, indeed much better. I have pushed.